### PR TITLE
Adds a package for LvArray

### PR DESCRIPTION
**Pull request by corbett5**
Loosens the restrictions on clang when using CUDA on a PowerPC system. These may not be officially supported but they work on the LLNL sierra systems.

Bugfix for CHAI when building with RAJA.
Added variants to both CHAI and RAJA to disable examples, benchmarks, and exercises.

Added a package for LvArray.

### CODE REVIEW

**Comment by corbett5** [comment]
I can't add reviewers but @adrienbernede @davidbeckingsale should take a look at the RAJA and CHAI changes.

**Comment by davidbeckingsale** [comment]
RAJA and CHAI changes look okay. @corbett5 - take a look at this branch for a way around the CUDA issue: https://github.com/davidbeckingsale/spack/tree/feature/allow-untested-cuda-versions

**Comment by corbett5** [comment]
> 
> 
> RAJA and CHAI changes look okay. @corbett5 - take a look at this branch for a way around the CUDA issue: https://github.com/davidbeckingsale/spack/tree/feature/allow-untested-cuda-versions

@davidbeckingsale nice! I like that a lot better.

**Comment by davidbeckingsale** [comment]
> > RAJA and CHAI changes look okay. @corbett5 - take a look at this branch for a way around the CUDA issue: https://github.com/davidbeckingsale/spack/tree/feature/allow-untested-cuda-versions
> 
> @davidbeckingsale nice! I like that a lot better.

Feel free to use the fork for now. I am going to turn it into a PR soon.

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (corbett5) [inline var/spack/repos/builtin/packages/lvarray/package.py:35]: In LvArray I have an uberenv based build setup. This involves duplicating this `package.py` file with some additional logic to generate a CMake host config file. Following the convention of the other projects that also use uberenv (RAJA, CHAI, Umpire, Axom) I have omitted the host config logic from this package. However is this necessary? If acceptable I'd prefer to add the host config logic here so I could get rid of the very similar `package.py` in LvArray.

https://github.com/GEOSX/LvArray/blob/temp/feature/corbett/tribol/scripts/uberenv/packages/lvarray/package.py

**Comment by corbett5** [comment]
> 
> Feel free to use the fork for now. I am going to turn it into a PR soon.

@davidbeckingsale I rebased on top of your branch. If you'd prefer not to include your commits in this PR I can wait until yours is merged and then I can drop them.

**Comment by corbett5** [comment]
The mpich build is failing because Spack can't find a Fortran compiler, but this test isn't required. Is this something that I need to fix?

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (adrienbernede) [reply var/spack/repos/builtin/packages/lvarray/package.py:35]: @corbett5 please put the host-config logic here. It’s perfectly fine from Spack perspective, and it’s what RAJA, CHAI and Umpire will do soon too).

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lvarray/package.py]:
```suggestion
                make('docs')
```

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lvarray/package.py]:
```suggestion
            ctest('-V', '--force-new-ctest-process')
```

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lvarray/package.py]: Probably `type='build'`

**Review by adamjstewart — COMMENTED** [review]
@svenevs @ax3l may want to comment on the CUDA changes.

**Comment by svenevs** [comment]
> Feel free to use the fork for now. I am going to turn it into a PR soon.

I'm not a huge fan of this, but if that's how things are going to go can we review that separately (putting this on hold until then)? The version compatibility stuff is already really difficult to maintain, the nvidia tables aren't even accurate in the docs so ax3l trolls the header files (thanks again).

https://gist.github.com/ax3l/9489132

Maybe the powerpc changes you have here are "supported" in the headers, in which case we should just add them. In other words, introducing the variant is what I'm opposed to, it either works or it doesn't and if it works let's just change it without needing to track an additional ~variant set of versions.

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (corbett5) [reply var/spack/repos/builtin/packages/lvarray/package.py]: Much, better thanks!

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (corbett5) [reply var/spack/repos/builtin/packages/lvarray/package.py]: Thanks!

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (corbett5) [reply var/spack/repos/builtin/packages/lvarray/package.py]: As a keyword argument?
```
depends_on('doxygen@1.8.13', when'+docs', type='build')`? 
```
And would this only add the dependency when executing the build stage?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/lvarray/package.py]: Yes, that's the correct syntax. The dependency would only be added to the `PATH` and `PYTHONPATH` during installation, it wouldn't be linked to using Spack's compiler wrappers or affect the hash in any way.

Comment (corbett5) [reply var/spack/repos/builtin/packages/lvarray/package.py]: Nice, done.

File: var/spack/repos/builtin/packages/lvarray/package.py
Comment (corbett5) [reply var/spack/repos/builtin/packages/lvarray/package.py:35]: Done.

**Comment by corbett5** [comment]
> 
> 
> > Feel free to use the fork for now. I am going to turn it into a PR soon.
> 
> I'm not a huge fan of this, but if that's how things are going to go can we review that separately (putting this on hold until then)? The version compatibility stuff is already really difficult to maintain, the nvidia tables aren't even accurate in the docs so ax3l trolls the header files (thanks again).
> 
> https://gist.github.com/ax3l/9489132
> 
> Maybe the powerpc changes you have here are "supported" in the headers, in which case we should just add them. In other words, introducing the variant is what I'm opposed to, it either works or it doesn't and if it works let's just change it without needing to track an additional ~variant set of versions.

@svenevs I'll happily make that change, as is I need to add `+allow-untested-versions` to every package that uses cuda in the spec.

File: lib/spack/spack/build_systems/cuda.py
Comment (corbett5) [inline lib/spack/spack/build_systems/cuda.py]: @svenevs these are the conflicts I could see happening on our system. They are just some of the conflicts @davidbeckingsale had modified with `allow-untested-versions`.

Comment (davidbeckingsale) [reply lib/spack/spack/build_systems/cuda.py]: Some versions are "unofficially" supported - these are what I wanted to allow with the new variant. We have modified CUDA installs at LLNL where the headers are modified to allow these combinations to work.

Comment (corbett5) [reply lib/spack/spack/build_systems/cuda.py]: Ahh I didn't realize we had modified headers.

Comment (svenevs) [reply lib/spack/spack/build_systems/cuda.py]: I think there's a better solution here,

- Introduce the variant for allowing unsupported. For CUDA 11+ this will require us to introduce the newly added build flag for this, and disables conflicts. (This solution won't help you).
    - This will also relate to the LLVM // CUDA work.
- Wherever and however that happens, sneak in a hack for the exact versions LLNL needs with a `NOTE` explaining this is the only exception we're going to add explaining LLNL gets special priority.

I do still think it would be better to open a separate PR for this, we can merge it without solving the CUDA 11 thing but instead open an issue to track that and reference the PR adding the variant to indicate it's already there but does not do anything yet. Basically, keep the history clean but additionally since CUDA 11 allows version checks to be suppressed, the variant will deceive people so the help string needs to explain it's not going to do what they think.

I'm not willing to sign off on this in it's current form because of the confusion it will cause ("I set allow unsupported but it doesn't work" influx of issues). But I've also stepped down from maintaining this package and think the current maintainers should have the opportunity to vet this on it's own without the other changes going on here.

Comment (corbett5) [reply lib/spack/spack/build_systems/cuda.py]: I've removed all CUDA changes from this branch including the commit history.

**Comment by ax3l** [comment]
> may want to comment on the CUDA changes.

looks quite package specific, but what I can spot from CMake-ish parts it LGTM

**Comment by corbett5** [comment]
Are there any outstanding issues that should be addressed?

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
